bgpd/bgp_nht.c

```diff
@@ -843,9 +843,17 @@ static void evaluate_paths(struct bgp_nexthop_cache *bnc)
 	if (peer) {
 		int valid_nexthops = bgp_isvalid_nexthop(bnc);
 
-		if (valid_nexthops)
-			peer->last_reset = PEER_DOWN_WAITING_OPEN;
-		else
+		if (valid_nexthops) {
+			/*
+			 * Peering cannot occur across a blackhole nexthop
+			 */
+			if (bnc->nexthop_num == 1
+			    && bnc->nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
```

Comment:
should this test just be encapsulated in bgp_isvalid_nexthop() - is it possible there are other paths where "valid" doesn't really mean "useable"?

Reply:
I think I agree here ... it seems like this should be part of `bgp_isvalid_nexthop`, so this condition will be caught in other places, rather than just here ...

Reply:
no because this is fundamentally only a peer thing.  We use the bgp_isvalid_nexthop for non-peer ( say static connected routes or just regular nexthop validation ) that we do want to consider that a blackhole route is valid to be used.

Reply:
The goal here is to stop the bgp fsm from allowing to try to connect through a blackhole route.  Which just dumps a million messages of unable to connect.

Reply:
@donaldsharp is right, it's not actually _nexthop_ thing, but for peering.

Reply:
Got it, thanks